Src/Base/AMReX_BoxIterator.H

```diff
@@ -12,152 +11,179 @@
 
 namespace amrex
 {
-  /**
-     \brief iterates through the IntVects of a Box
-
-
-     BoxIterator iterates through the IntVects of a box.  The actual
-     sequence of IntVects is implementation-specific.
-     Typical usage:
-
-     Box b;
-     ...
-     BoxIterator bit (b);
-     for (bit.begin(); bit.ok(); ++bit)
-     {
-     IntVect iv = bit();
-     (do operations involving iv)
-     }
-  */
-  class BoxIterator
-  {
-  public:
-    ///
+
+template<int dim>
+class BoxND;
+
+/**
+* \brief iterates through the IntVects of a Box
+*
+* BoxIteratorND iterates through the IntVects of a box
+* in the same order as a series of nested for loops
+* with the innermost loop iterating over the first dimension.
+*
+* Typical usage using a range-based for loop:
+* \code{.cpp}
+*   Box b;
+*   ...
+*   for (IntVect iv : b.iterator()) {
+*       // do operations involving iv
+*   }
+* \endcode
+*
+* Using a traditional for loop:
+* \code{.cpp}
+*   Box b;
+*   ...
+*   BoxIterator bit (b);
+*   for (bit.begin(); bit.ok(); ++bit)
+*   {
+*       IntVect iv = bit();
+*       // do operations involving iv
+*   }
+* \endcode
+*
+* Note that while regular iteration of the range-based version is reasonably performant,
+* neither version is compatible with omp parallel for or simd autovectorization.
+*/
+template<int dim>
+class BoxIteratorND
+{
+public:
+
     /**
-       Default constructor.  This constructs an invalid iterator.
-       The user must call define before using.
+    * Default constructor. This constructs an invalid iterator.
+    * The user must call define before using.
     */
-    BoxIterator () noexcept = default;
+    BoxIteratorND () noexcept = default;
 
-    ///
     /**
-       Constructs a BoxIterator and associates it with a Box.
-       Arguments:
-       a_bx (not modified) the Box to iterate over.
+    * Constructs a BoxIteratorND and associates it with a Box.
+    * Arguments:
+    * a_bx (not modified) the Box to iterate over.
     */
-    explicit BoxIterator (const Box& a_bx) noexcept;
+    explicit BoxIteratorND (const BoxND<dim>& a_bx) noexcept {
+        define(a_bx);
+    }
 
-    void setBox (const Box& a_bx) noexcept;
+    void setBox (const BoxND<dim>& a_bx) noexcept {
+        define(a_bx);
+    }
+
+    /**
+    * Associates a Box with this BoxIteratorND.
+    * Arguments:
+    * a_bx (not modified) the Box to iterate over.
+    */
+    void define (const BoxND<dim>& a_bx) noexcept {
+        if (a_bx.ok() && a_bx.smallEnd().allLE(a_bx.bigEnd())) {
```

Comment:
The second part of the test seems redundant.

Reply:
It is. It was a leftover from the previous version. I removed it now.